File: codechef/merge.cpp
#include<bits/stdc++.h>
using namespace std;


void merge(int arr[],int left,int right){
    int mid=(left+right)/2;
    int lsize=mid-left+1;
    int rsize=right-mid;
    int pos=left;
    int leftarr[lsize];
    int rightarr[rsize];
    for(int i=0;i<lsize;i++){
        leftarr[i]=arr[left+i];
    }

    for(int i=0;i<rsize;i++){
        rightarr[i]=arr[mid+1+i];
    }

    int i=0,j=0;
    while(i<lsize && j<rsize){
        if(leftarr[i]<=rightarr[j]){
            arr[pos++]=leftarr[i++];
        }
        else{
            arr[pos++]=rightarr[j++];
        }
    }

    while(i<lsize){
        arr[pos++]=leftarr[i++];
    }
    while(j<rsize){
        arr[pos++]=rightarr[j++];
    }
}
void merge_sort(int arr[],int low,int high){
    if(low==high) return;
    int med=low+high/2;
    cout<<"hey";
    merge_sort(arr,low,med);
    merge_sort(arr,med+1,high);
    merge(arr,low,high);
}
int main(){
    int arr[]={4,54,46,9,5,12,53,21,76};
    merge_sort(arr,0,8);
    for(int i=0;i<9;i++){
        cout<<arr[i]<<"\t";
    }
}

File: codechef/merge.py
def merge(a,b):
    a_len=len(a)
    b_len=len(b)
    array=[]
    i=0
    j=0
    while(i<a_len and j<b_len):
        if(a[i]>b[j]):
            array.append(b[j])
            j+=1
        else:
            array.append(a[i])
            i+=1
    while i<a_len:
        array.append(a[i])
        i+=1
    while j<b_len:
        array.append(b[j])
        j+=1
    return array
def merge_sort(arr):

    arr_len=len(arr)
    if(arr_len==1):return arr
    median=arr_len//2
    left_arr=merge_sort(arr[:median])
    right_arr=merge_sort(arr[median:])
    return merge(left_arr,right_arr)
    
a=[4,56,12,6,23,7,8]
array=merge_sort(a)
for i in array:
    print(i, end=' ')